src/bpf/bpfd-manager.c

```diff
@@ -0,0 +1,448 @@
+                }
+
+                /* Fetch bpf_map_info using fd */
+                r = sym_bpf_obj_get_info_by_fd(fd, &map_info, &map_info_len);
+                if (r < 0) {
+                        log_error("Can't get bpf map info for fd: %i", fd);
+                        continue;
+                }
+
+                /* Store map names into hashmap for later lookup when iterating progs */
+                map_name = strdup(map_info.name);
+                if (!map_name)
+                        return -ENOMEM;
+                hashmap_put(map_name_by_id, UINT32_TO_PTR(map_id), map_name);
+
+                /* Fetch memlock from /proc/self/fdinfo using map fd */
+                r = get_memlock_by_fd(fd, &memlock);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to read map memlock from /proc/self/fdinfo/%i", fd);
+
+                if (link) {
+                        r = json_build(&v, JSON_BUILD_OBJECT(
+                                                JSON_BUILD_PAIR_STRING("TYPE", "MAP")));
+
+                        r = json_variant_merge_objectb(&v, JSON_BUILD_OBJECT(
+                                                JSON_BUILD_PAIR_OBJECT("DATA",
+                                                JSON_BUILD_PAIR_STRING("NAME", map_info.name),
+                                                JSON_BUILD_PAIR_UNSIGNED("TYPE", map_info.type),
+                                                JSON_BUILD_PAIR_STRING("MEMLOCK", memlock))));
+
+                        varlink_notify(link, v);
+                } else {
+                        log_struct(LOG_INFO,
+                                LOG_MESSAGE("Logging info about BPF map: %s", map_info.name),
+                                "MESSAGE_ID=" SD_MESSAGE_BPFD_LOG_STR,
+                                "NAME=%s", map_info.name,
+                                "TYPE=%u", map_info.type,
+                                "MEMLOCK=%s", memlock);
+                }
+        }
+
+        /* Iterate through all attached BPF progs*/
+        id = 0;
+        while (sym_bpf_prog_get_next_id(id, &id) >= 0) {
+                _cleanup_close_ int fd = -EBADF;
+                _cleanup_free_ char *memlock = NULL;
+                _cleanup_free_ uint32_t *prog_map_ids = NULL;
+                _cleanup_free_ char *prog_name = NULL;
+                _cleanup_free_ char *prog_map_names = NULL;
+                JsonVariant *v = NULL;
+
+                zero(prog_info);
+                /* Grab the FDs using their IDs*/
+                fd = sym_bpf_prog_get_fd_by_id(id);
+                if (fd < 0) {
+                        log_error("Can't get fd for prog id: %u", id);
+                        continue;
+                }
+
+                /* Fetch bpf_prog_info using fd */
+                r = sym_bpf_obj_get_info_by_fd(fd, &prog_info, &prog_info_len);
+                if (r < 0) {
+                        log_error("Can't get bpf prog info for fd: %i", fd);
```

Comment:
this returns a useful error code, doesn't it? please use log_error_errno() in that case, and include error string via `%m` in output